### PR TITLE
[dvc] Decompress/assemble record before transformation

### PR DESCRIPTION
**Pull request by kvargha**
## Summary

Events consumed from the version topic may be compressed/chunked. So we need to reassemble them before they're transformed.

## How was this PR tested?
CI.

## Does this PR introduce any user-facing changes?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

**Comment by kvargha** [comment]
Hey @ZacAttack, I'm facing this error "Could not deserialize bytes back into Avro object". Do you have an idea what's causing it?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 1]
```diff
@@ -3080,12 +3084,40 @@ private int processKafkaDataMessage(
         // update checksum for this PUT message if needed.
         partitionConsumptionState.maybeUpdateExpectedChecksum(keyBytes, put);
 
-        // Do transorfmation recompute key, value and partition
+        ByteBuffer valueBytes = put.getPutValue();
+        int putSchemaId = put.getSchemaId();
+        Schema valueSchema = schemaRepository.getValueSchema(storeName, putSchemaId).getSchema();
+
+        // Decompress/assemble record
```
Comment (ZacAttack) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3091]: We should only be doing any of this if the record transformer is enabled, so we should move this logic into the if block (if (recordTransformer != null)

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 2]
```diff
@@ -3080,12 +3083,33 @@ private int processKafkaDataMessage(
...
+              keyBytes,
+              valueBytes,
+              consumerRecord.getOffset(),
+              Lazy.of(() -> new AvroGenericDeserializer<>(valueSchema, valueSchema)),
+              putSchemaId,
+              compressorFactory.getCompressor(compressionStrategy, kafkaVersionTopic));
+
+          // Current record is a chunk. We only write to the storage engine for fully assembled records
+          if (assembledObject == null) {
+            break;
```
Comment (ZacAttack) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3107]: I don't think is the right thing to do here, I think return; would be good.

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3107]: What value should I return here, since the return type for this function is an int?

Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3107]: I think it'd be fine to return zero since the return value is what was committed to storage (we didn't write anything to storage at this stage).

**Review by ZacAttack — APPROVED** [review]
This overall looks good.  Let's see if we can add a test though that triggers the chunking behavior just to be certain.  That or enable compression or both.

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientTest.java [hunk 3]
```diff
@@ -204,6 +204,13 @@ public void testBatchStore(DaVinciConfig clientConfig) throws Exception {
     MetricsRepository metricsRepository = new MetricsRepository();
 
     // Test record transformation
+
+    // Turn on GZIP compression
+    cluster.useControllerClient(controllerClient -> {
```
Comment (ZacAttack) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientTest.java:209]: So this is gonna get noOp compression because compression will only get enabled on the next version of the store. 

Follow this call up with a cluster.getNewVersion() call.  This will force the store to take the new copmression config.

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientTest.java [hunk 4]
```diff
@@ -229,6 +236,13 @@ public void testBatchStore(DaVinciConfig clientConfig) throws Exception {
       clientConfig.setRecordTransformer(null);
     }
 
+    // Turn off GZIP compression
```
Comment (ZacAttack) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientTest.java:239]: I don't think this is necessary, we can leave it on I think.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/VersionBackend.java [hunk 5]
```diff
@@ -95,8 +101,15 @@ public class VersionBackend {
     this.stopConsumptionTimeoutInSeconds =
         backend.getConfigLoader().getCombinedProperties().getInt(SERVER_STOP_CONSUMPTION_TIMEOUT_IN_SECONDS, 60);
     this.storeDeserializerCache = backend.getStoreOrThrow(store.getName()).getStoreDeserializerCache();
-    this.compressor = Lazy.of(
-        () -> backend.getCompressorFactory().getCompressor(version.getCompressionStrategy(), version.kafkaTopicName()));
+
+    // If recordTransformer is enabled, we need to disable compression since it is already done by the recordTransformer
+    if (recordTransformer != null) {
```
Comment (ZacAttack) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/VersionBackend.java:106]: Why is this necessary?  Does this mean that decompression was being done earlier?

Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/VersionBackend.java:106]: Oh wait, actually I suppose this makes sense.  Davinci would be storing the data compressed and decompressing on read I think.  Do I have it right?

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/VersionBackend.java:106]: I faced an issue in an earlier commit where the data already was decompressed at an earlier stage with record transformation on, but I can't replicate it anymore. Let me push a change without this to ensure that it passes every single CI stage.

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/VersionBackend.java]: It's possible that when the data was written the version was still in NoOp and by the time it tried to read compression was set to GZIP.

**Comment by kvargha** [comment]
It looks like it's failing at diffCoverage. 

I added a test to StoreIngestionTaskTest but it's still saying it's not covered.

Here's what I have:
`@Test
  public void testRecordTransformer() {
    Version version = mock(Version.class);
    doReturn(1).when(version).getPartitionCount();
    doReturn(VersionStatus.STARTED).when(version).getStatus();
    doReturn(true).when(version).isNativeReplicationEnabled();

    Store store = mock(Store.class);
    doReturn(Optional.of(version)).when(store).getVersion(eq(1));

    VeniceStoreVersionConfig storeConfig = mock(VeniceStoreVersionConfig.class);
    doReturn(topic).when(storeConfig).getStoreVersionName();

    StoreIngestionTaskFactory ingestionTaskFactory = getIngestionTaskFactoryBuilder(
        new RandomPollStrategy(),
        Utils.setOf(PARTITION_FOO),
        Optional.empty(),
        1,
        Collections.emptyMap(),
        true).build();
    
    TestRecordTransformer recordTransformer = new TestRecordTransformer();
    storeIngestionTaskUnderTest = ingestionTaskFactory.getNewIngestionTask(
        store,
        version,
        new Properties(),
        isCurrentVersion,
        storeConfig,
        1,
        false,
        Optional.empty(),
        recordTransformer);

    VenicePartitioner partitioner = getVenicePartitioner(1);
    int targetPartitionPutKeyFoo = partitioner.getPartitionId(putKeyFoo, PARTITION_COUNT);
    mockAbstractStorageEngine.put(targetPartitionPutKeyFoo, ingestionTaskFactory, putValue); 
  }`

**Comment by elijahgrimaldi** [comment]
> It looks like it's failing at diffCoverage.
> 
> I added a test to StoreIngestionTaskTest but it's still saying it's not covered. Any idea what's wrong?
> 
> Here's what I have:
> 
> ```
> public class TestRecordTransformer
>     implements DaVinciRecordTransformer<Integer, String, TransformedRecord<Integer, String>> {
>   public Schema getKeyOutputSchema() {
>     return Schema.create(Schema.Type.INT);
>   }
> 
>   public Schema getValueOutputSchema() {
>     return Schema.create(Schema.Type.STRING);
>   }
> 
>   public TransformedRecord<Integer, String> put(Lazy<Integer> key, Lazy<String> value) {
>     TransformedRecord<Integer, String> transformedRecord = new TransformedRecord<>();
>     transformedRecord.setKey(key.get());
>     transformedRecord.setValue(value.get() + "Transformed");
>     return transformedRecord;
>   }
> }
> 
> @Test
> public void testRecordTransformer() {
>   Version version = mock(Version.class);
>   doReturn(1).when(version).getPartitionCount();
>   doReturn(VersionStatus.STARTED).when(version).getStatus();
>   doReturn(true).when(version).isNativeReplicationEnabled();
> 
>   Store store = mock(Store.class);
>   doReturn(Optional.of(version)).when(store).getVersion(eq(1));
> 
>   VeniceStoreVersionConfig storeConfig = mock(VeniceStoreVersionConfig.class);
>   doReturn(topic).when(storeConfig).getStoreVersionName();
> 
>   StoreIngestionTaskFactory ingestionTaskFactory = getIngestionTaskFactoryBuilder(
>       new RandomPollStrategy(),
>       Utils.setOf(PARTITION_FOO),
>       Optional.empty(),
>       1,
>       Collections.emptyMap(),
>       true).build();
>   
>   TestRecordTransformer recordTransformer = new TestRecordTransformer();
>   storeIngestionTaskUnderTest = ingestionTaskFactory.getNewIngestionTask(
>       store,
>       version,
>       new Properties(),
>       isCurrentVersion,
>       storeConfig,
>       1,
>       false,
>       Optional.empty(),
>       recordTransformer);
> 
>   VenicePartitioner partitioner = getVenicePartitioner(1);
>   int targetPartitionPutKeyFoo = partitioner.getPartitionId(putKeyFoo, PARTITION_COUNT);
>   mockAbstractStorageEngine.put(targetPartitionPutKeyFoo, ingestionTaskFactory, putValue); 
> }
> ```

Could we import `TestRecordTransformer` to avoid duplicate code? And possibly rename the test to differentiate between this and the test in `RecordTransformerTest`.

**Comment by adamxchen** [comment]
Hey @elijahgrimaldi , for the current failure, from the report, it looks like these lines are not covered.
Screenshot 2024-01-28 at 1 41 01 PM

From your unit test above, it does look like you are passing a valid transformer, but did your codes invoke `processKafkaDataMessage`, where your new codes are added, implicitly? If not, this path is not exercised by your unit test. You can add some breakpoints in your IDE to find out.

**Comment by kvargha** [comment]
> Hey @kvargha , for the current failure, from the report, it looks like these lines are not covered. Screenshot 2024-01-28 at 1 41 01 PM
> 
> From your unit test above, it does look like you are passing a valid transformer, but did your codes invoke `processKafkaDataMessage`, where your new codes are added, implicitly? If not, this path is not exercised by your unit test. You can add some breakpoints in your IDE to find out.

Hi @adamxchen, thanks for chiming in. `processKafkaDataMessage` isn't being invoked, and I'm having trouble figuring out how to get it to execute without directly calling that method.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 6]
```diff
@@ -3104,15 +3107,37 @@ private int processKafkaDataMessage(
...
 
-        // Do transorfmation recompute key, value and partition
+        // Check if put.getSchemaId is positive, if not default to 1
+        int putSchemaId = put.getSchemaId() > 0 ? put.getSchemaId() : 1;
+
+        // Do transformation recompute key, value and partition
         if (recordTransformer != null) {
           long recordTransformStartTime = System.currentTimeMillis();
+          ByteBuffer valueBytes = put.getPutValue();
+          Schema valueSchema = schemaRepository.getValueSchema(storeName, putSchemaId).getSchema();
```
Comment (kvargha) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: I got the unit test to reach here but schemaRepository.getValueSchema returns null. How would I set the schema from within the unit test?

Comment (adamxchen) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: You can set up the schemaRepo in `StoreIngestionTaskFactory` via the builder pattern

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: It looks like schemaRepo is being set in `getIngestionTaskFactoryBuilder`. How would I define the schema after schemaRepo has already been setup?

When I log `schemaRepository.getValueSchemas(storeName)` it returns an empty list.

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: This has been resolved in my latest commit.

Comment (adamxchen) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: The spotbug failed tho. Can you take a look?

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: The CI has this error:
```
Execution failed for task ':clients:da-vinci-client:diffCoverage'.
> java.lang.Exception: Rule violated for bundle da-vinci-client: branches covered ratio is 0.3, but expected minimum is 0.4
```
But when I run it locally, it passes. Do you have an idea why?

image

Comment (adamxchen) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: It's very strange that JDK17 passes but JDK 11 and 8 failed. What's the JDK version on your local?

Comment (kvargha) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: I'm using JDK17 locally.

Comment (adamxchen) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: This is very strange.. I pulled down your PR and ran it locally using JDK11. It passed w/o issue. Let us re-run the jobs.. In the meanwhile, please work on other open comments and maybe add a bit more tests to see if it helps??

Comment (adamxchen) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:3117]: Okay, the retry worked

**Review by ZacAttack — APPROVED** [review]
Looks good!  Thanks for all the effort!
